File: spiral-matrix/spiral-matrix.js
var spiralOrder = function(matrix) {
    if (matrix == null || matrix.length === 0) return [];

  const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];// right, down, left, up
  const steps = [matrix[0].length, matrix.length - 1];
    // console.log( matrix[0]);
  let dir = 0;
  let x = 0;
  let y = -1;
  const res = [];

  while (steps[dir % 2] > 0) {
    for (let i = 0; i < steps[dir % 2]; i++) {
      x += dirs[dir][0];
      y += dirs[dir][1];
      res.push(matrix[x][y]);
    }
    steps[dir % 2]--;
     // console.log(steps);
    dir = (dir + 1) % 4;
  }
  return res;

};

File: spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {

        int dir =0;
        int length = matrix.length;
        int width = matrix[0].length;
        int up =0, down =length-1, right = width-1, left = 0;
        int r =0, c = 0;
        List<Integer> list = new ArrayList<>();
        while(list.size() < length*width){

            if(dir == 0){
                r = up;
                for(int i=left; i<=right; i++){
                    list.add(matrix[r][i]);
                }
                up++;
                dir=1;
            }
            else if(dir == 1){
                c = right;
                for(int i=up; i<=down; i++){
                    list.add(matrix[i][c]);
                }
                right--;
                dir = 2;
            }
            else if(dir == 2){
                r = down;
                for(int i=right; i>=left; i--){
                    list.add(matrix[r][i]);
                }
                down--;
                dir = 3;
            }
            else if(dir == 3){
                c = left;
                for(int i=down; i>=up; i--){
                    list.add(matrix[i][c]);
                }
                left++;
                dir=0;
            }
        }
        return list;
    }
}